### PR TITLE
improve agent/server telemetry

### PR DESCRIPTION
**Pull request by azdagron**
- added additional telemetry to agent
- introduced and added telemetry to server

### CODE REVIEW

File: pkg/agent/endpoints/workload/handler.go [hunk 1]
```diff
@@ -49,13 +49,14 @@ func (h *Handler) FetchJWTSVID(ctx context.Context, req *workload.JWTSVIDRequest
...
 	}
 
-	_, selectors, _, done, err := h.startCall(ctx)
+	_, selectors, metrics, done, err := h.startCall(ctx)
 	if err != nil {
 		return nil, err
 	}
 	defer done()
 
 	// TODO: telemetry
```
Comment (evan2645) [inline pkg/agent/endpoints/workload/handler.go]: Is this comment stale now?

File: pkg/agent/endpoints/workload/handler.go [hunk 2]
```diff
@@ -129,15 +130,30 @@ func (h *Handler) ValidateJWTSVID(ctx context.Context, req *workload.ValidateJWT
...
 
-	metrics.IncrCounter([]string{workloadApi, "validate_jwt"}, 1)
-
 	keyStore := keyStoreFromBundles(h.getWorkloadBundles(selectors))
 
 	spiffeID, claims, err := jwtsvid.ValidateToken(ctx, req.Svid, keyStore, []string{req.Audience})
 	if err != nil {
+		metrics.IncrCounterWithLabels([]string{workloadApi, "validate_jwt_svid"}, 1, []telemetry.Label{
+			{
+				Name:  "error",
```
Comment (evan2645) [inline pkg/agent/endpoints/workload/handler.go]: Errors are usually modeled as a separate counter, e.g. "spire.validate_jwt_svid" and "spire.validate_jwt_svid.error".

Including the error message itself can lead to problems related to label cardinality, and AFAICT will make it more difficult to see overall error count.

File: pkg/agent/svid/rotator.go [hunk 3]
```diff
@@ -84,7 +85,11 @@ func (r *rotator) shouldRotate() bool {
 }
 
 // rotateSVID asks SPIRE's server for a new agent's SVID.
-func (r *rotator) rotateSVID(ctx context.Context) error {
+func (r *rotator) rotateSVID(ctx context.Context) (err error) {
+	counter := telemetry.StartCall(r.c.Metrics, &err, "svid", "rotate")
```
Comment (evan2645) [inline pkg/agent/svid/rotator.go]: nit: perhaps this should be `agent_svid` to distinguish it from workload svid rotation without having to refer to a label value

File: pkg/common/telemetry/call.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	err     *error
+	key     []string
+	labels  []Label
+	start   time.Time
+}
+
+func StartCall(metrics Metrics, err *error, key string, keyn ...string) *CallCounter {
+	return &CallCounter{
+		metrics: metrics,
+		err:     err,
```
Comment (evan2645) [inline pkg/common/telemetry/call.go]: Is it safe to give this error pointer directly to the telemetry package? You would have to be sure not to update error after having called Done()...?

Perhaps there are some other options for how to handle an error condition, considering my comment above about error counter emission?

File: pkg/common/telemetry/metrics.go [hunk 5]
```diff
@@ -48,7 +48,8 @@ var _ Metrics = (*MetricsImpl)(nil)
 // NewMetrics returns a Metric implementation
 func NewMetrics(c *MetricsConfig) *MetricsImpl {
 	// Always create an in-memory sink
-	interval := 1 * time.Second
+	interval := 1 * time.Minute
+	//interval := 1 * time.Second
```
Comment (evan2645) [inline pkg/common/telemetry/metrics.go]: Was this change intentional? IMO 1 second is a reasonable emission frequency for runtime-related statistics

File: pkg/common/telemetry/metrics.go [hunk 6]
```diff
@@ -59,7 +60,12 @@ func NewMetrics(c *MetricsConfig) *MetricsImpl {
 	// error and the implementation is currently no-fail.
 	sinks := metrics.FanoutSink{inmemSink}
 	sinks = append(sinks, c.Sinks...)
-	m, _ := metrics.New(metrics.DefaultConfig(c.ServiceName), sinks)
+
+	conf := metrics.DefaultConfig(c.ServiceName)
+	conf.EnableHostname = true
+	conf.EnableHostnameLabel = true
+	//conf.ProfileInterval = time.Minute
```
Comment (evan2645) [inline pkg/common/telemetry/metrics.go]: Was this intentional?

File: pkg/common/telemetry/metrics.go [hunk 7]
```diff
@@ -59,7 +60,12 @@ func NewMetrics(c *MetricsConfig) *MetricsImpl {
 	// error and the implementation is currently no-fail.
 	sinks := metrics.FanoutSink{inmemSink}
 	sinks = append(sinks, c.Sinks...)
-	m, _ := metrics.New(metrics.DefaultConfig(c.ServiceName), sinks)
+
+	conf := metrics.DefaultConfig(c.ServiceName)
+	conf.EnableHostname = true
```
Comment (evan2645) [inline pkg/common/telemetry/metrics.go]: I think we should prefer adding hostname as a label and not as a part of the metric name... systems which don't support labels can apply logic in the sink itself to inject labels into the metric name, that way systems which _do_ support labels aren't hindered

File: pkg/server/ca/ca.go [hunk 8]
```diff
@@ -97,7 +99,14 @@ func (ca *serverCA) SignX509SVID(ctx context.Context, csrDER []byte, ttl time.Du
 		return nil, err
 	}
 
-	ca.c.Log.Debugf("Signed x509 SVID %q (expires %s)", cert.URIs[0].String(), cert.NotAfter.Format(time.RFC3339))
+	spiffeID := cert.URIs[0].String()
+	ca.c.Log.Debugf("Signed x509 SVID %q (expires %s)", spiffeID, cert.NotAfter.Format(time.RFC3339))
+	ca.c.Metrics.IncrCounterWithLabels([]string{"server_ca", "sign", "x509_svid"}, 1, []telemetry.Label{
```
Comment (evan2645) [inline pkg/server/ca/ca.go]: are these counters emitted with `spire_server` prefix, or `spire.server.` or something along those lines? We should probably do that if we're not, in which case this can probably go from `server_ca` to just `ca`?

File: pkg/server/ca/ca.go [hunk 9]
```diff
@@ -131,5 +140,20 @@ func (ca *serverCA) SignJWTSVID(ctx context.Context, jsr *node.JSR) (string, err
...
+
+	labels := []telemetry.Label{
+		{
+			Name:  "spiffe_id",
+			Value: jsr.SpiffeId,
+		},
+	}
+	for _, audience := range jsr.Audience {
+		labels = append(labels, telemetry.Label{
+			Name:  "audience",
```
Comment (evan2645) [inline pkg/server/ca/ca.go]: can we have multiple labels with the same name?

File: pkg/server/svid/rotator.go [hunk 10]
```diff
@@ -88,7 +89,8 @@ func (r *rotator) shouldRotate() bool {
 
 // rotateSVID cuts a new server SVID from the CA plugin and installs
 // it on the endpoints struct. Also updates the CA certificates.
-func (r *rotator) rotateSVID(ctx context.Context) error {
+func (r *rotator) rotateSVID(ctx context.Context) (err error) {
+	defer telemetry.CountCall(r.c.Metrics, &err, "rotate", "svid")()
```
Comment (evan2645) [inline pkg/server/svid/rotator.go]: The agent has `"svid", "rotate"`

File: pkg/common/telemetry/metrics.go (hunk 5 above)
Comment (azdagron) [reply pkg/common/telemetry/metrics.go]: It was not. Will revert.

File: pkg/common/telemetry/metrics.go (hunk 7 above)
Comment (azdagron) [reply pkg/common/telemetry/metrics.go]: Done.

File: pkg/common/telemetry/metrics.go (hunk 6 above)
Comment (azdagron) [reply pkg/common/telemetry/metrics.go]: Nope! Reverted.

File: pkg/server/ca/ca.go (hunk 8 above)
Comment (azdagron) [reply pkg/server/ca/ca.go]: The service name (i.e "spire_server" or "spire_agent") is prefixed to keys unless you set EnableServiceLabel to `true`. The default value (which we don't change) is false.

File: pkg/server/ca/ca.go (hunk 9 above)
Comment (azdagron) [reply pkg/server/ca/ca.go]: Metrics package itself doesn't do any processing of labels outside of filtering... so i guess it would be up to the sink?

File: pkg/server/svid/rotator.go (hunk 10 above)
Comment (azdagron) [reply pkg/server/svid/rotator.go]: good catch.

File: pkg/server/ca/ca.go (hunk 8 above)
Comment (evan2645) [reply pkg/server/ca/ca.go]: Awesome... thoughts on `server_ca` -> `ca`?

File: pkg/server/ca/ca.go (hunk 9 above)
Comment (evan2645) [reply pkg/server/ca/ca.go]: Ok. suggest we test this stuff out with datadog or something similar before cutting a release with it
